private/bufpkg/bufcheck/buflint/internal/buflintcheck/buflintcheck.go

```diff
@@ -963,3 +968,95 @@ func checkSyntaxSpecified(add addFunc, file protosource.File) error {
 	}
 	return nil
 }
+
+var (
+	// fieldConstraintsDesc provides a Descriptor for validate.FieldConstraints.
+	fieldConstraintsDesc = (*validate.FieldConstraints)(nil).ProtoReflect().Descriptor()
+
+	// fieldConstraintsOneOfDesc provides the OneOfDescriptor for the type union
+	// in FieldConstraints.
+	fieldConstraintsOneOfDesc = fieldConstraintsDesc.Oneofs().ByName("type")
+)
+
+// CheckValidateRulesTypesMatch is a check function.
+var CheckValidateRulesTypesMatch = newMessageCheckFunc(checkValidateRulesTypesMatch)
+
+func checkValidateRulesTypesMatch(add addFunc, message protosource.Message) error {
+	for _, field := range message.Fields() {
+		data, err := getDataByExtension(field, validate.E_Field)
+		if err != nil {
+			return err
+		}
+		var constraints validate.FieldConstraints
+		if err := proto.Unmarshal(data, &constraints); err != nil {
+			return fmt.Errorf("unmarshal error for field %q: %v", field.FullName(), err)
+		}
+		constraintsRefl := constraints.ProtoReflect()
+		setOneOf := constraintsRefl.WhichOneof(fieldConstraintsOneOfDesc)
+		if setOneOf == nil {
+			continue
+		}
+
+		got := field.Type()
+		expected := getTypeName(setOneOf.Name())
```

Comment:
This is where the function name gets confusing - since this isn't the type name, and is instead a `FieldDescriptorProto_Type`, I think this needs a bit more explanation. This is relying on the fact that the field names in protovalidate happen to line up with `FieldDescriptorProto_Types` - which is fine if we are spelling out that this is our assumption, but I think this needs some code comments and better variable naming if possible.